[althea28] iP

althea28 opened the pull request:
# **DUCK  🦆🦆**
> __"HONK! HONK! HONK!"__ _-Duck_

Duck helps you track everything you need to do.
It's:
- Fun
- Easy
- Fast

To use it:
1. download from [here](https://github.com/althea28/ip)
2. double-click it to start chatting
3. add tasks
4. let **DUCK** manage your tasks for you

**Features:**
- [x] Managing tasks
- [ ] Honking
- [ ] 🐣

Here is the first few lines of the `main` method:
```ruby
public static void main(String[] args) throws DuckException {
        Scanner scan = new Scanner(System.in);
        TaskList cmds = new TaskList(Storage.load(), Storage.loadNum());
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

althea28 committed:
Level-0

althea28 committed:
Level-1

althea28 committed:
Level-2

althea28 committed:
Level-3

althea28 committed:
Level-4

althea28 committed:
A-TextUiTesting

althea28 committed:
Level-5

althea28 committed:
Level-6

althea28 committed:
no message

althea28 committed:
branch-Level-7

althea28 committed:
no message

althea28 committed:
Merge branch 'master' of ssh://github.com/althea28/ip

* 'master' of ssh://github.com/althea28/ip:
  branch-Level-7

althea28 committed:
no message

althea28 committed:
no message

althea28 committed:
Merge branch 'master' of ssh://github.com/althea28/ip

* 'master' of ssh://github.com/althea28/ip:
  no message
  no message
  branch-Level-7
  no message

althea28 committed:
no message

althea28 committed:
no message

althea28 committed:
Merge branch 'master' of ssh://github.com/althea28/ip

* 'master' of ssh://github.com/althea28/ip:
  no message

althea28 committed:
no message

althea28 committed:
no message

althea28 committed:
Merge branch 'master' of ssh://github.com/althea28/ip

* 'master' of ssh://github.com/althea28/ip:
  no message
  no message
  no message
  no message
  no message
  no message
  branch-Level-7
  no message

althea28 committed:
Level-6-reset

althea28 committed:
level-7 fix

althea28 committed:
level-7 test

althea28 committed:
Merge branch 'master' into branch-Level-7

* master:
  level-7 test

althea28 committed:
no message

althea28 committed:
Add dates and times recognition

althea28 committed:
Do test-8 branching test

althea28 committed:
Merge branch 'master' into branch-Level-8

* master:
  Do test-8 branching test

althea28 committed:
Add OOP to code structure

althea28 committed:
Clean OOP code

althea28 committed:
Divide classes into packages

althea28 committed:
Merge branch 'master' into add-gradle-support

* master: (26 commits)
  Divide classes into packages
  Clean OOP code
  Add OOP to code structure
  Do test-8 branching test
  Add dates and times recognition
  no message
  level-7 test
  level-7 fix
  Level-6-reset
  no message
  no message
  no message
  no message
  no message
  no message
  no message
  branch-Level-7
  no message
  Level-6
  Level-5
  ...

althea28 committed:
Add Gradle Support

althea28 committed:
Add JUnit Tests

althea28 committed:
Add JAR files

althea28 committed:
Add Java Documentations for all classes

althea28 committed:
Add Java coding conventions

althea28 committed:
Add find function

althea28 committed:
Merge branch 'master' into branch-A-CodingStandard

* master:
  Add Java Documentations for all classes

# Conflicts:
#	src/main/java/duke/Parser.java

althea28 committed:
Merge branch 'master' into branch-Level-9

* master:
  Add Java coding conventions
  Add Java Documentations for all classes

# Conflicts:
#	src/main/java/duke/Duck.java
#	src/main/java/duke/Parser.java
#	src/main/java/duke/Storage.java
#	src/main/java/duke/TaskList.java

althea28 committed:
Update build.gradle

althea28 committed:
Add GUI

althea28 committed:
Delete unnecessary classes

xinweichong commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,147 @@
+package duke;
+
+import java.io.File;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
```
Perhaps you can separate the imports into 3 blocks with ```java.io```, ```java.nio``` and ```java.util``` for better visibility and organisation

althea28 replied:
Noted, will do so!

xinweichong commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,147 @@
+        cmdNum = n;
+    }
+
+    /**
+     * Adds new task to list.
+     *
+     * @param t Task to be added to the list.
+     * */
```
Perhaps you can add ```@return``` to your Javadoc to provide some description on what this method is trying to return (noticed the same point throughout the code)

althea28 replied:
Yep, forgot to update some of the JavaDocs after editing the code from Week 3. Thanks for the catch!

xinweichong commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,75 @@
+                    userCmd.substring(start + 1 + end + 4)));
+
+        } else {
+            return "I don't recognise that command.";
```
Perhaps this can be thrown as a ```DuckException``` instead of returning it as a String!

xinweichong commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,147 @@
+        } else if (task instanceof Event) {
+            type = "E";
+        }
+        return type+" | "+done+" | "+task.description + task.getDates();
```
Perhaps you could separate the operators and variable names for better readability!

althea28 replied:
Oops, thanks for the catch, will ammend accordingly!

xinweichong commented on src/main/java/duke/DialogBox.java:
```diff
@@ -0,0 +1,59 @@
+import javafx.scene.layout.HBox;
+
+/**
+ * Represents a dialog box consisting of an ImageView to represent the speaker's face
+ * and a label containing text from the speaker.
```
Great descriptions given in Javadoc throughout the entire project - clear and aids in understanding of the code!

yanhwee commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,70 @@
+ */
+public class Task {
+    protected String description;
+    protected boolean isDone;
+    public Task(String description) {
```
Perhaps a new line between field and method declaration would look be more readable?

yanhwee commented on src/main/java/duke/DuckException.java:
```diff
@@ -0,0 +1,10 @@
+ * Represents exception encoutered by Duck chatbot.
+ */
+public class DuckException extends Exception {
+    //tester 1
```
```suggestion
```

yanhwee commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,32 @@
+ * Represents a Task with a specified start and end time.
+ */
+public class Event extends Task{
+    LocalDate fromdate;
+    LocalDate todate;
```
```suggestion
    LocalDate fromDate;
    LocalDate toDate;
```

variable names should be in camelCase?

yanhwee commented on src/main/java/duke/MainWindow.java:
```diff
@@ -0,0 +1,51 @@
+
+    private Duck duck;
+
+    private Image userImage = new Image(this.getClass().getResourceAsStream("/images/user.png"));
+    private Image dukeImage = new Image(this.getClass().getResourceAsStream("/images/duck.png"));
```
File paths preferably extracted out as constant variables?

yanhwee commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,147 @@
+            if (!file.exists()) {
+                return 0;
+            } else {
+                List<String> tasklist = Files.readAllLines(Paths.get(PATH));
```
```suggestion
                List<String> taskList = Files.readAllLines(Paths.get(PATH));
```

yanhwee commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,147 @@
+            done = "1";
+        }
+        String type = "";
+        if (task instanceof Todo) {
+            type = "T";
+        } else if (task instanceof Deadline) {
+            type = "D";
+        } else if (task instanceof Event) {
+            type = "E";
+        }
```
This seems like applying polymorphism can help!
i.e. adding a `public char saveTaskType()` to each of the `Todo`, `Deadline`, `Event` task classes.
But anyways, I feel that this is rather minor.

yanhwee commented on src/test/java/duke/TaskListTest.java:
```diff
@@ -0,0 +1,19 @@
+
+public class TaskListTest {
+    @Test
+    public void test() throws DuckException {
```
More descriptive test names would be preferable?
E.g. `unitBeingTested_descriptionOfTestInputs_expectedOutcome`

yanhwee reviewed (commented):
Looks good to me! Just some nits to fix. Furthermore, you might want to package your files, for instance you might want to put all your task-related classes (e.g. `Deadline`, `Todo`, `Event`) into a single package so it's easier to traverse the directory of files.
👍👍

althea28 committed:
Amend javadocs

Minor javadocs updates due to code structure change.

althea28 committed:
Add assertions

althea28 committed:
Refactor code

Long methods and complicated expressions are used to parse through each user command in the Parser class.

Long methods and complicated expressions make it hard for the reader to understand the coding logic.

Breaking down long methods and complicated expressions into smaller methods make their purposes clearer and more understandable for the reader.

Let's break down the matching and execution of each command into different methods.

This makes it easier for the reader to recognise how each command is recognised and matched, as well as how each command is executed.

althea28 committed:
Update javadocs

Currently no javadocs are added for the new methods created after the code was refactored.

The lack of javadocs may make it unclear to the reader the purpose and function of each method.

Adding javadocs makes it easier for reader to understand the code logic as it improves the readability of the code.

Let's add javadocs for the purpose, parameters and return values for all new methods created.

Adding javadocs makes it easier to understand all existing code.

althea28 committed:
Merge pull request #1 from althea28/branch-A-Assertions

Add assertions

althea28 committed:
Merge branch 'master' into branch-A-CodeQuality

* master:
  Add assertions

# Conflicts:
#	src/main/java/duke/Duck.java
#	src/main/java/duke/Parser.java
#	src/main/java/duke/Storage.java
#	src/main/java/duke/TaskList.java

althea28 committed:
Merge pull request #2 from althea28/branch-A-CodeQuality

Refactor code

althea28 committed:
Add Undo extension

Currently no undo function call for Duck.

This makes it inconvenient for the user to reverse their last command, as they will have to manually reverse it themselves.

Adding an undo extension makes it easier for users to undo their previous command, in cases where they made a mistake or change their mind.

Let's add an Undo extension by using a new class Undo, to track the previous command and reverse it if called.

Using a new class to handle this undo feature prevents the logic from getting confusing and messy, if it were all stored in the Parser class instead.

althea28 committed:
Remove dead code

In some cases commented out code was left in codebase.

Removing dead code improves readability and tidiness.

Let's remove all commented out code from all classes in the codebase. Doing so will make the code tidier.

althea28 committed:
Handle invalid task numbers in user input

parseCommand() throws exception when user inputs task numbers that are 0 or negative.

Handling these cases informs user of invalid input so that the user understands the error in their command.

Let's handle these cases by returning a string informing the user when an invalid task number is detected.

althea28 committed:
Add ParserTest

althea28 committed:
Add UndoTest

althea28 committed:
Improve code quality

althea28 committed:
Add Ui.png

althea28 committed:
Update README.md

althea28 committed:
Delete duke.txt file
